### PR TITLE
Replace nl-unit-test with pigweed for protocols/secure_channel

### PR DESCRIPTION
**Pull request by jlatusek**
## Problem
The problem was described in #29342. 

## Changes
These changes are another piece of replacing the nl-unit test with the pigweed-unit test. I changed tests for the `protocols/secure_channel`.

### CODE REVIEW

File: src/lib/support/UnitTestPigweedUtils.h [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
Comment (arkq) [inline src/lib/support/UnitTestPigweedUtils.h:1]: Add license.

File: config/pw_unit_test/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import("//build_overrides/build.gni")
```
Comment (arkq) [inline config/pw_unit_test/BUILD.gn:1]: License header

File: src/protocols/secure_channel/tests/TestPairingSession.cpp [hunk 3]
```diff
@@ -36,9 +36,17 @@
...
 
-class TestPairingSession : public PairingSession
+class TestPairingSession : public PairingSession, public ::testing::Test
 {
 public:
+    static void SetUpTestSuite()
+    {
+
+        CHIP_ERROR error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
```
Comment (arkq) [inline src/protocols/secure_channel/tests/TestPairingSession.cpp:46]: ASSERT

File: src/protocols/secure_channel/tests/TestStatusReport.cpp [hunk 4]
```diff
@@ -19,19 +19,29 @@
...
 using namespace chip::Protocols;
 using namespace chip::Protocols::SecureChannel;
 
-void TestStatusReport_NoData(nlTestSuite * inSuite, void * inContext)
+struct TestStatusReport : public ::testing::Test
+{
+    static void SetUpTestSuite()
+    {
+        CHIP_ERROR error = chip::Platform::MemoryInit();
+        EXPECT_EQ(error, CHIP_NO_ERROR);
```
Comment (arkq) [inline src/protocols/secure_channel/tests/TestStatusReport.cpp:39]: ASSERT

File: src/protocols/secure_channel/tests/TestPASESession.cpp [hunk 5]
```diff
@@ -288,31 +282,27 @@ void SecurePairingHandshakeTestCommon(nlTestSuite * inSuite, void * inContext, S
...
-        auto waitTimeout = 100_ms;
+        auto waitTimeout = 100_ms + CHIP_CONFIG_MRP_RETRY_INTERVAL_SENDER_BOOST;
 
-#if CHIP_CONFIG_ENABLE_ICD_SERVER
-        // If running as an ICD, increase waitTimeout to account for:
-        // - longer MRP intervals, configured above to 1s/1s,
-        // - the fast-polling interval that is added to the MRP backoff time.
-        waitTimeout += 2000_ms32;
-        waitTimeout += ICDConfigurationData::GetInstance().GetFastPollingInterval();
+#if CHIP_CONFIG_ENABLE_ICD_SERVER == 1
```
Comment (arkq) [inline src/protocols/secure_channel/tests/TestPASESession.cpp:303]: Why this change?

**Review by arkq — APPROVED** [review]

File: src/protocols/secure_channel/tests/TestCASESession.cpp [hunk 6]
```diff
@@ -1028,231 +991,172 @@ void TestCASESession::SessionResumptionStorage(nlTestSuite * inSuite, void * inC
...
+        EXPECT_EQ(establishmentReturnVal, CHIP_NO_ERROR);
+        EXPECT_EQ(loopback.mSentMessageCount, testVectors[i].expectedSentMessageCount);
+        EXPECT_EQ(delegateCommissioner.mNumPairingComplete, i + 1);
         SessionHolder & holder = delegateCommissioner.GetSessionHolder();
-        NL_TEST_ASSERT(inSuite, bool(holder));
-        NL_TEST_ASSERT(inSuite, holder->GetPeer() == fabricInfo->GetScopedNodeIdForNode(Node01_01));
+        EXPECT_TRUE(bool(holder));
+        EXPECT_EQ(holder->GetPeer(), fabricInfo->GetScopedNodeIdForNode(Node01_01));
         chip::Platform::Delete(pairingCommissioner);
+        gPairingServer.Shutdown();
```
Comment (jlatusek) [inline src/protocols/secure_channel/tests/TestCASESession.cpp:1017]: I've added this line in the test to fix the error in the next test, `Sigma1BadDestinationIdTest`. Without this modification, the SIGABRT signal was received during call `DrainAndServiceIO` in `Sigma1BadDestinationIdTest`

File: src/lib/support/UnitTestPigweedUtils.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+/*
+ * Run fixture's class function as a test.
```
Comment (andy31415) [inline src/lib/support/UnitTestPigweedUtils.h:21]: Could you add an example of usage in this documentation?

File: src/protocols/secure_channel/tests/TestCASESession.cpp [hunk 8]
```diff
@@ -26,6 +26,8 @@
 #include <credentials/PersistentStorageOpCertStore.h>
 #include <crypto/DefaultSessionKeystore.h>
 #include <errno.h>
+#include <gtest/gtest.h>
+#include <ios>
```
Comment (andy31415) [inline src/protocols/secure_channel/tests/TestCASESession.cpp:30]: Do we need the ios header? Generally we try to avoid that on things that may run on embedded. It seems not needed before.

File: src/protocols/secure_channel/tests/TestCASESession.cpp [hunk 9]
```diff
@@ -334,22 +340,25 @@ CHIP_ERROR InitCredentialSets()
...
-    VerifyOrDieWithMsg((err = chip::DeviceLayer::PlatformMgr().InitChipStack()) == CHIP_NO_ERROR, AppServer,
-                       "Init CHIP stack failed: %" CHIP_ERROR_FORMAT, err.Format());
-    VerifyOrDieWithMsg((err = InitFabricTable(gCommissionerFabrics, &gCommissionerStorageDelegate, /* opKeyStore = */ nullptr,
-                                              &gCommissionerOpCertStore)) == CHIP_NO_ERROR,
-                       AppServer, "InitFabricTable failed: %" CHIP_ERROR_FORMAT, err.Format());
-    VerifyOrDieWithMsg((err = InitCredentialSets()) == CHIP_NO_ERROR, AppServer, "InitCredentialSets failed: %" CHIP_ERROR_FORMAT,
-                       err.Format());
+
+    err = chip::DeviceLayer::PlatformMgr().InitChipStack();
+    ASSERT_EQ(err, CHIP_NO_ERROR) << "Init CHIP stack failed: " << std::hex << err.Format();
```
Comment (andy31415) [inline src/protocols/secure_channel/tests/TestCASESession.cpp:349]: we can drop the extra messages bits ... I think if we include StringBuilderAdapters (after https://github.com/project-chip/connectedhomeip/pull/33505) we can do:

```cpp
ASSERT_EQ(chip::DeviceLayer::PlatformMgr().InitChipStack(), CHIP_NO_ERROR);
```

And you will get a nicely formatter info on failure. The only purpose of these messages seem to have been for having the CHIP-error nicely formatted and our macros can do that already.

Comment (arkq) [reply src/protocols/secure_channel/tests/TestCASESession.cpp:349]: Also, one thing to add. Generally it's always better to do:
```c++
ASSERT_EQ(chip::DeviceLayer::PlatformMgr().InitChipStack(), CHIP_NO_ERROR);
```
instead of:
```c++
CHIP_ERROR err = chip::DeviceLayer::PlatformMgr().InitChipStack();
ASSERT_EQ(err, CHIP_NO_ERROR);
```
In the former case, the test framework will be able to print the name of the function which has failed instead of a variable name, which might not say much (then you will have to look at the test source code to know which function has failed).

File: src/lib/support/BUILD.gn [hunk 10]
```diff
@@ -341,6 +341,15 @@ static_library("test_utils") {
   ]
 }
 
+pw_source_set("pw_test_utils") {
+  output_dir = "${root_out_dir}/lib"
+  public_deps = [
+    "$dir_pw_log:impl",
+    "$dir_pw_unit_test",
+  ]
+  sources = [ "UnitTestPigweedUtils.h" ]
```
[the remark this answers is not in the document]
Comment (jlatusek) [reply src/lib/support/BUILD.gn]: Fixed

File: src/lib/support/UnitTestPigweedUtils.h (hunk 7 above)
Comment (jlatusek) [reply src/lib/support/UnitTestPigweedUtils.h]: Fixed

File: src/protocols/secure_channel/tests/TestCASESession.cpp (hunk 8 above)
Comment (jlatusek) [reply src/protocols/secure_channel/tests/TestCASESession.cpp]: Fixed

File: src/protocols/secure_channel/tests/TestCASESession.cpp (hunk 9 above)
Comment (jlatusek) [reply src/protocols/secure_channel/tests/TestCASESession.cpp]: Fixed

**Review by arkq — APPROVED** [review]

File: src/protocols/secure_channel/tests/TestStatusReport.cpp [hunk 11]
```diff
@@ -73,121 +82,57 @@ void TestStatusReport_WithData(nlTestSuite * inSuite, void * inContext)
...
 
     const System::PacketBufferHandle & rcvData = reportToParse.GetProtocolData();
-    if (rcvData.IsNull())
-    {
-        NL_TEST_ASSERT(inSuite, false);
-        return;
-    }
-    NL_TEST_ASSERT(inSuite, rcvData->DataLength() == dataLen);
-    NL_TEST_ASSERT(inSuite, !memcmp(rcvData->Start(), data, dataLen));
+    EXPECT_FALSE(rcvData.IsNull());
```
Comment (feasel0) [inline src/protocols/secure_channel/tests/TestStatusReport.cpp:94]:
```suggestion
    ASSERT_FALSE(rcvData.IsNull());
```

File: src/protocols/secure_channel/tests/TestStatusReport.cpp [hunk 12]
```diff
@@ -73,121 +82,57 @@ void TestStatusReport_WithData(nlTestSuite * inSuite, void * inContext)
...
-    NL_TEST_ASSERT(inSuite, reportToParse.GetProtocolCode() == protocolCode);
+    EXPECT_EQ(reportToParse.Parse(std::move(handle)), CHIP_NO_ERROR);
+    EXPECT_EQ(reportToParse.GetGeneralCode(), generalCode);
+    EXPECT_EQ(reportToParse.GetProtocolId(), protocolId);
+    EXPECT_EQ(reportToParse.GetProtocolCode(), protocolCode);
 
     const System::PacketBufferHandle & rcvData = reportToParse.GetProtocolData();
-    NL_TEST_ASSERT(inSuite, !rcvData.IsNull());
-    NL_TEST_ASSERT(inSuite, rcvData->DataLength() == sizeof(minimumWaitTime));
+    EXPECT_FALSE(rcvData.IsNull());
```
Comment (feasel0) [inline src/protocols/secure_channel/tests/TestStatusReport.cpp:129]: Maybe ASSERT here since we're dereferencing rcvData on the next line.

File: src/protocols/secure_channel/tests/TestPASESession.cpp [hunk 13]
```diff
@@ -323,244 +317,188 @@ void SecurePairingHandshakeTestCommon(nlTestSuite * inSuite, void * inContext, S
...
     // Standalone acks also increment the mSentMessageCount. But some messages could be acked
     // via piggybacked acks. So we cannot check for a specific value of mSentMessageCount.
     // Let's make sure atleast number is >= than the minimum messages required to complete the
     // handshake.
-    NL_TEST_ASSERT(inSuite, loopback.mSentMessageCount >= sTestPaseMessageCount);
-    NL_TEST_ASSERT(inSuite, delegateAccessory.mNumPairingErrors == 0);
-    NL_TEST_ASSERT(inSuite, delegateAccessory.mNumPairingComplete == 1);
-    NL_TEST_ASSERT(inSuite, delegateCommissioner.mNumPairingErrors == 0);
-    NL_TEST_ASSERT(inSuite, delegateCommissioner.mNumPairingComplete == 1);
+    EXPECT_TRUE(loopback.mSentMessageCount >= sTestPaseMessageCount);
```
Comment (feasel0) [inline src/protocols/secure_channel/tests/TestPASESession.cpp:328]:
```suggestion
    EXPECT_GE(loopback.mSentMessageCount, sTestPaseMessageCount);
```

File: src/protocols/secure_channel/tests/TestCASESession.cpp [hunk 14]
```diff
@@ -982,19 +939,19 @@ void TestCASESession::SessionResumptionStorage(nlTestSuite * inSuite, void * inC
 
     // Create our fabric-scoped node IDs.
     const FabricInfo * fabricInfo = gCommissionerFabrics.FindFabricWithIndex(gCommissionerFabricIndex);
-    NL_TEST_ASSERT(inSuite, fabricInfo != nullptr);
+    EXPECT_NE(fabricInfo, nullptr);
```
Comment (feasel0) [inline src/protocols/secure_channel/tests/TestCASESession.cpp:942]:
```suggestion
    ASSERT_NE(fabricInfo, nullptr);
```
Since fabricInfo is about to be dereferenced

File: src/protocols/secure_channel/tests/TestCASESession.cpp [hunk 15]
```diff
@@ -551,100 +518,93 @@ void TestCASESession::SecurePairingHandshakeServerTest(nlTestSuite * inSuite, vo
...
+    EXPECT_EQ(loopback.mSentMessageCount, sTestCaseMessageCount);
+    EXPECT_EQ(delegateCommissioner.mNumPairingComplete, 1u);
 
     // Validate that secure session is created
     SessionHolder & holder = delegateCommissioner.GetSessionHolder();
-    NL_TEST_ASSERT(inSuite, bool(holder));
+    EXPECT_TRUE(bool(holder));
 
-    NL_TEST_ASSERT(inSuite, (holder->GetPeer() == chip::ScopedNodeId{ Node01_01, gCommissionerFabricIndex }));
+    EXPECT_TRUE((holder->GetPeer() == chip::ScopedNodeId{ Node01_01, gCommissionerFabricIndex }));
```
Comment (feasel0) [inline src/protocols/secure_channel/tests/TestCASESession.cpp:545]: Can this be done with EXPECT_EQ?

File: src/protocols/secure_channel/tests/TestStatusReport.cpp (hunk 12 above)
Comment (feasel0) [inline src/protocols/secure_channel/tests/TestStatusReport.cpp:129]:
```suggestion
    ASSERT_FALSE(rcvData.IsNull());
```
Since we're about to dereference rcvData

**Review by feasel0 — APPROVED** [review]

File: src/protocols/secure_channel/tests/TestCASESession.cpp (hunk 6 above)
Comment (jlatusek) [reply src/protocols/secure_channel/tests/TestCASESession.cpp:1017]: @andy31415 Is it fine? Or should I fix it in another way?

**Review by andy31415 — APPROVED** [review]
